### PR TITLE
feat(apigateway): feature flag to disable the default endpoint Output

### PR DESCRIPTION
**Pull request by jamesorlakin**
The `RestApi` construct creates a CfnOutput by default with no exportName. This seems slightly peculiar, and in stacks of my own I've often resorted to `myApi.node.tryRemoveChild('Endpoint')` to avoid confusion when the deployed APIs are private (and need accessing via a VPC endpoint instead of the normally public HTTPS URL).

It was contemplated in https://github.com/aws/aws-cdk/issues/3662#issuecomment-521432190 to remove this entirely. With CDK v2 providing an opportunity to allow breaking behaviours I've created a new feature flag to gradually introduce this new behaviour. I appreciate breaking changes are never desired, so I'm not disheartened if this is unwanted (feel free to close!).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts [hunk 1]
```diff
@@ -551,7 +554,12 @@ export abstract class RestApiBase extends Resource implements IRestApi {
         ...props.deployOptions,
       });
 
-      new CfnOutput(this, 'Endpoint', { exportName: props.endpointExportName, value: this.urlForPath() });
+      // If the context is true, don't create an unnamed output.
+      // If false - or the endpointExportName is set, do create a CfnOutput.
+      const avoidEmptyExport: boolean | undefined = Node.of(this).tryGetContext(APIGATEWAY_REST_API_NO_CFN_OUTPUT);
```
Comment (nija-at) [inline packages/@aws-cdk/aws-apigateway/lib/restapi.ts:559]: I believe you should use the `FeatureFlags` API - https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/core/lib/feature-flags.ts#L11 - instead.

File: packages/@aws-cdk/cx-api/lib/features.ts [hunk 2]
```diff
@@ -160,6 +160,15 @@ export const LAMBDA_RECOGNIZE_VERSION_PROPS = '@aws-cdk/aws-lambda:recognizeVers
  */
 export const CLOUDFRONT_DEFAULT_SECURITY_POLICY_TLS_V1_2_2021 = '@aws-cdk/aws-cloudfront:defaultSecurityPolicyTLSv1.2_2021';
 
+/**
+ * API Gateway's REST API construct always creates a CfnOutput with a blank exportName. That name can be overridden
```
Comment (nija-at) [inline packages/@aws-cdk/cx-api/lib/features.ts:164]:
```suggestion
 * API Gateway's RestApi construct always creates a CfnOutput with a blank exportName. That name can be overridden
```

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for taking the effort to do this.
This is excellent!

Just a couple of small comments.

File: packages/@aws-cdk/aws-apigateway/lib/restapi.ts (hunk 1 above)
Comment (jamesorlakin) [reply packages/@aws-cdk/aws-apigateway/lib/restapi.ts]: Ooh, nice spot!

**Comment by jamesorlakin** [comment]
That's an excellent question! I'll be honest and say I didn't think about doing that 🤦‍♂️ 

An extra boolean prop like `avoidEndpointExport` could work, then it could throw if provided a `endpointExportName` as well. While the default of having an output seems a little odd I guess it's not a breaking change by any stretch.

**Comment by nija-at** [comment]
@jamesorlakin - feels to me like the simpler solution than introducing a feature flag.

I can see you've put in a considerable amount of work to get the feature flag going, especially given that this is a tricky part of our code base, but would you consider switching up the solution?

**Review by nija-at — CHANGES_REQUESTED** [review]
@jamesorlakin - feels to me like the simpler solution than introducing a feature flag.

I can see you've put in a considerable amount of work to get the feature flag going, especially given that this is a tricky part of our code base, but would you consider switching up the solution?

**Comment by jamesorlakin** [comment]
Hey @nija-at, that's honestly fine and no worries - good thinking of a much simpler solution!

I'll close this one and maybe make something similar to the `deploy` prop (default true) when I get around to it.

**Comment by olka-dev** [comment]
@jamesorlakin you have not got around it in end, haven`t you?...
You could save my day with this feature flag. 
Unfortunately you were stopped before merging, @nija-at big thanks there... 🤬

**Comment by jamesorlakin** [comment]
...nope, it fell off my radar. You can always open a PR if you're that passionate about it. 👀 

Otherwise a one-liner `tryRemoveChild` is far from the end of the world. Subclass with this behaviour if there's many RestApi instances and you don't want repetition in other constructs.

**Pull request closed without merging**